playback/exoplayer/src/main/kotlin/ExoPlayerBackend.kt

```diff
@@ -0,0 +1,109 @@
+package org.jellyfin.playback.exoplayer
+
+import android.content.Context
+import com.google.android.exoplayer2.C
+import com.google.android.exoplayer2.DefaultRenderersFactory
+import com.google.android.exoplayer2.ExoPlayer
+import com.google.android.exoplayer2.MediaItem
+import com.google.android.exoplayer2.PlaybackException
+import com.google.android.exoplayer2.Player
+import com.google.android.exoplayer2.video.VideoSize
+import org.jellyfin.playback.core.backend.BasePlayerBackend
+import org.jellyfin.playback.core.mediastream.MediaStream
+import org.jellyfin.playback.core.model.PlayState
+import org.jellyfin.playback.core.model.PositionInfo
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.ZERO
+import kotlin.time.Duration.Companion.milliseconds
+
+class ExoPlayerBackend(
+	private val context: Context,
+) : BasePlayerBackend() {
+	override fun getPositionInfo(): PositionInfo = PositionInfo(
+		active = exoPlayer.currentPosition.milliseconds,
+		buffer = exoPlayer.bufferedPosition.milliseconds,
+		duration = if (exoPlayer.duration == C.TIME_UNSET) ZERO else exoPlayer.duration.milliseconds,
+	)
+
+	private val exoPlayer by lazy {
+		val renderersFactory = DefaultRenderersFactory(context).apply {
+			setEnableDecoderFallback(true)
+			setExtensionRendererMode(DefaultRenderersFactory.EXTENSION_RENDERER_MODE_ON)
+		}
+		ExoPlayer.Builder(context, renderersFactory)
+			.build()
+			.also { player -> player.addListener(PlayerListener()) }
+	}
+
+	inner class PlayerListener : Player.Listener {
+		override fun onIsPlayingChanged(isPlaying: Boolean) {
+			val state = when {
+				isPlaying -> PlayState.PLAYING
+				exoPlayer.playbackState == Player.STATE_IDLE || exoPlayer.playbackState == Player.STATE_ENDED -> PlayState.STOPPED
+				else -> PlayState.PAUSED
+			}
+			listener?.onPlayStateChange(state)
+		}
+
+		override fun onPlayerError(error: PlaybackException) {
+			listener?.onPlayStateChange(PlayState.ERROR)
+		}
+
+		override fun onVideoSizeChanged(size: VideoSize) {
+			listener?.onVideoSizeChange(size.width, size.height)
+		}
+
+		override fun onPlaybackStateChanged(playbackState: Int) {
+			onIsPlayingChanged(exoPlayer.isPlaying)
+		}
+	}
+
+	override fun prepareStream(stream: MediaStream) {
+		val mediaItem = MediaItem.Builder().apply {
+			setMediaId(stream.hashCode().toString())
+			setUri(stream.url)
+		}.build()
+
+		if (exoPlayer.mediaItemCount == 0) {
+			exoPlayer.setMediaItem(mediaItem)
+		} else {
+			// Remove old items
+			while (exoPlayer.mediaItemCount > 1) exoPlayer.removeMediaItem(0)
+			// Add new item
+			exoPlayer.addMediaItem(mediaItem)
+		}
```

Comment:
Is there a benefit to this method over just doing

```
while (exoPlayer.mediaItemCount > 1) exoPlayer.removeMediaItem(0)
exoPlayer.addMediaItem(mediaItem)

Reply:
Nop, fixed it